File: rust/core/src/extension.rs
use anyhow::Result;
use std::collections::HashMap;
use tokio::sync::RwLock;
use tokio_util::sync::CancellationToken;

use crate::{
    client_data::ExtensionClient,
    data::{
        action::{EventData, EventResult},
        activity::EntryActivity,
        extension_repo::{ExtensionRepo, RemoteExtensionResult},
        settings::Setting,
        source::{
            EntryDetailed, EntryDetailedResult, EntryId, EntryList, EpisodeId, Source, SourceResult,
        },
    },
    store::ExtensionStore,
};

#[async_trait::async_trait()]
pub trait Adapter: Send + Sync {
    async fn get_extensions(&self) -> Result<Vec<Box<dyn Extension>>>;

    async fn install(&self, url: &str) -> Result<Box<dyn Extension>>;
    async fn uninstall(&self, ext: &Box<dyn Extension>) -> Result<()>;

    async fn get_repo(&self, url: &str) -> Result<ExtensionRepo>;
    async fn browse_repo(&self, repo: &ExtensionRepo, page: i32) -> Result<RemoteExtensionResult>;
}

#[async_trait::async_trait()]
pub trait Extension: Send + Sync {
    fn is_enabled(&self) -> bool;
    fn get_data(&self) -> &RwLock<ExtensionStore>;
    fn get_client(&self) -> &dyn ExtensionClient;
    async fn set_enabled(&mut self, enabled: bool) -> Result<()>;
    async fn reload(&mut self) -> Result<()>;

    async fn event(
        &self,
        event: EventResult,
        token: Option<CancellationToken>,
    ) -> Result<Option<EventData>>;

    async fn browse(&self, page: i32, token: Option<CancellationToken>) -> Result<EntryList>;

    async fn search(
        &self,
        page: i32,
        filter: String,
        token: Option<CancellationToken>,
    ) -> Result<EntryList>;

    async fn handle_url(&self, url: String, token: Option<CancellationToken>) -> Result<bool>;

    async fn detail(
        &self,
        entryid: EntryId,
        settings: HashMap<String, Setting>,
        token: Option<CancellationToken>,
    ) -> Result<EntryDetailedResult>;

    async fn source(
        &self,
        epid: EpisodeId,
        settings: HashMap<String, Setting>,
        token: Option<CancellationToken>,
    ) -> Result<SourceResult>;

    async fn map_entry(
        &self,
        entry: EntryDetailed,
        settings: HashMap<String, Setting>,
        token: Option<CancellationToken>,
    ) -> Result<EntryDetailedResult>;

    async fn on_entry_activity(
        &self,
        activity: EntryActivity,
        entry: EntryDetailed,
        settings: HashMap<String, Setting>,
        token: Option<CancellationToken>,
    ) -> Result<()>;

    async fn map_source(
        &self,
        source: Source,
        epid: EpisodeId,
        settings: HashMap<String, Setting>,
        token: Option<CancellationToken>,
    ) -> Result<SourceResult>;
}

File: tests/dion/arguments/extension.ts
/// <reference path="../../../js/dion_extension_types/index.d.ts" />
import { assert, DefaultExtension } from "../extensionutils.js";

export default class extends DefaultExtension {
  browse(page: number): Promise<EntryList> {
    assert(typeof page === "number", "Argument page of browse is not a number");
    return super.browse(page);
  }
  detail(
    entryid: EntryId,
    settings: Record<string, Setting>,
  ): Promise<EntryDetailedResult> {
    assert(
      typeof entryid === "object" && typeof entryid.uid === "string",
      "Argument entryid of detail is not a string",
    );
    assert(
      typeof settings === "object" && settings !== null,
      "Argument settings of detail is not an object",
    );
    return super.detail(entryid, settings);
  }

  handleUrl(url: string): Promise<boolean> {
    assert(typeof url === "string", "Argument url of fromUrl is not a string");
    return super.handleUrl(url);
  }

  mapEntry(
    entry: EntryDetailed,
    settings: Record<string, Setting>,
  ): Promise<EntryDetailedResult> {
    assert(
      typeof entry === "object" && entry !== null,
      "Argument entry of mapEntry is not an object",
    );
    assert(
      typeof settings === "object" && settings !== null,
      "Argument settings of mapEntry is not an object",
    );
    return super.mapEntry(entry, settings);
  }

  mapSource(
    source: Source,
    epid: EpisodeId,
    settings: Record<string, Setting>,
  ): Promise<SourceResult> {
    assert(
      typeof source === "object" && source !== null,
      "Argument source of mapSource is not an object",
    );
    assert(
      typeof epid === "object" && typeof epid.uid === "string",
      "Argument epid of source is not a strin string",
    );
    assert(
      typeof settings === "object" && settings !== null,
      "Argument settings of mapSource is not an object",
    );
    return super.mapSource(source, epid, settings);
  }

  onEntryActivity(
    activity: EntryActivity,
    entry: EntryDetailed,
    settings: Record<string, Setting>,
  ): Promise<void> {
    assert(
      typeof activity === "object" && activity !== null,
      "Argument activity of onEntryActivity is not an object",
    );
    assert(
      typeof entry === "object" && entry !== null,
      "Argument entry of onEntryActivity is not an object",
    );
    assert(
      typeof settings === "object" && settings !== null,
      "Argument settings of onEntryActivity is not an object",
    );
    return super.onEntryActivity(activity, entry, settings);
  }

  search(page: number, filter: string): Promise<EntryList> {
    assert(typeof page === "number", "Argument page of search is not a number");
    assert(
      typeof filter === "string",
      "Argument filter of search is not a string",
    );
    return super.search(page, filter);
  }

  source(
    epid: EpisodeId,
    settings: Record<string, Setting>,
  ): Promise<SourceResult> {
    assert(
      typeof epid === "object" && typeof epid.uid === "string",
      "Argument epid of source is not a strin string",
    );
    assert(
      typeof settings === "object" && settings !== null,
      "Argument settings of source is not an object",
    );
    return super.source(epid, settings);
  }
}
